### PR TITLE
Code Review 1 for goyalbl

### PR DESCRIPTION
**Pull request by bipalgoyal**
**Analysis of the program.**
RedCord is an application where one can create user accounts, and after account creation, they can create posts, search for posts, send messages, and like and dislike posts within site.

**Was the program available in UC Github on time?**
Yes, I had enough time to complete the code review for this project.

**Is the program easy for you, as an outsider, to understand?**
Yes, it was straightforward to envision the application in my head after looking at the interactive storyboard and the detailed class diagram of the project. 

**Does the program compile?**
Yes, the program does indeed compile.

I have proposed changes for 4 different things, which I will mention here and then describe in detail if needed inside the file changes:

#### 1. Fixed the WhiteSpace in all of the code files
This helps improve the readability of the code, and will thus reduce the technical debt on future people, as code often changes hands in the real world.

**Source(s):** 
https://rhamedy.medium.com/a-short-summary-of-java-coding-best-practices-31283d0167d3 
(look at sections concerning Methods and If-checks)
https://www.baeldung.com/java-clean-code
(section 5.4 Whitespaces)

#### 2. Modified a Unit Test to follow given, when, then (Arrange, Act, and Assert) syntax:
This is in line with best practices 
**Source(s):** 
https://www.youtube.com/watch?v=Mv_yqVK_1ak 
(Video in Module 4 lecture where we see a hands-on example of TDD)
https://phauer.com/2019/modern-best-practices-testing-java/
(General section)

### 3. Added  Exception in fetchAllMessages in IMessageDAO and throw it all the way down to the Controller
This allows us to return the appropriate status code to our endpoint.
**Source(s):**
https://stackoverflow.com/questions/29114797/is-it-good-practice-to-throw-exception-from-dao-layer-to-controller
(The question asker brings up this point)

### 4. Added logger to RedCordController
When applications get more and more complex, logging becomes very useful in debugging, and getting it set up from the early stages will make things very easier in the long run

**Source(s):**
https://martinfowler.com/bliki/TechnicalDebt.html
https://martinfowler.com/articles/microservices.html

### CODE REVIEW

File: src/main/java/com/enterprise/redcord/RedCordController.java [hunk 1]
```diff
@@ -36,17 +36,17 @@ public String index(Model model) {
     public String saveTopic(Topic topic) {
        try {
            topicService.save(topic);
-       }catch(Exception e){
+       } catch(Exception e) {
            e.printStackTrace();
-        }
-        return "start";
```
Comment (bipalgoyal) [inline src/main/java/com/enterprise/redcord/RedCordController.java:42]: 1. Fixed the WhiteSpace

File: src/main/java/com/enterprise/redcord/RedCordController.java [hunk 2]
```diff
@@ -26,27 +30,35 @@
     @RequestMapping("/")
     public String index(Model model) {
         Message messageEntry = new Message();
+
+        logger.trace("Accessed index method in RedCordController.");
```
Comment (discospiff) [inline src/main/java/com/enterprise/redcord/RedCordController.java:34]: Excellent use of logging.

File: src/main/java/com/enterprise/redcord/RedCordController.java [hunk 3]
```diff
@@ -26,27 +30,35 @@
...
     
     @RequestMapping("/saveTopic")
     public String saveTopic(Topic topic) {
+       logger.trace("Accessed savedTopic method in RedCordController.");
+
        try {
            topicService.save(topic);
-       }catch(Exception e){
+       } catch(Exception e) {
+           logger.error("Error in saveTopic method: " + e.getMessage());
```
Comment (discospiff) [inline src/main/java/com/enterprise/redcord/RedCordController.java:48]: Great.  Logging is good here.  And you can remove the printStackTrace(), as you can get the same information (and then some) from the logger.

File: src/test/java/com/enterprise/redcord/RedCordApplicationTests.java [hunk 4]
```diff
@@ -57,25 +62,27 @@ private void thenCreateNewMessageRecordAndReturnIt() throws Exception {
...
 
-
     @Test
-    void save_fetch() throws Exception  {
-        saveTopic();
-        assertSaved();
+    void save_validateReturnTopicWithTopicIDUserIDTitleAndMessage() throws Exception {
+        givenTopicDataAreAvailable();
+        whenUserCreatesOrUpdatesTopicAndSaves();
+        thenCreateNewTopicRecordAndReturnIt();
```
Comment (discospiff) [inline src/test/java/com/enterprise/redcord/RedCordApplicationTests.java:69]: Good

**Comment by discospiff** [comment]
Several good suggestions here that reduce technical debt.  Logging is a great touch.  Using Given When Then is good syntax.  I recommend merging this branch.

**Pull request merged**